# Dataset

## Extract text from PDF

In [309]:
import os
import re
import json
import pandas as pd
import spacy
from PyPDF2 import PdfReader
import pdfplumber
from pathlib import Path
import logging
from typing import Dict, List, Any, Optional, Tuple, Set

In [310]:
nlp = spacy.load("en_core_web_lg")

## Get Data

- viewreport - good
- sample_report
- report 1
- report 2

In [1259]:
pdf_path = "reports/report 1.pdf"

In [1494]:
idx_section = 3

### Read PDF

In [1495]:
with open(pdf_path,"rb") as f:
    reader = PdfReader(f)
    text2 = ""
    for page in reader.pages:
        text2 += page.extract_text() + "\n"
print(text2[:400])


District 4, 27 - WINDSOR County
Owner: 1 - State Highway Agency
Town: 158 - NORWICH
Maintenance Responsibility: 1 - State Highway Agency
Team Lead:
 Stephen Piro,  
Inspection Date:
 04/15/2024
Structure 
#0048N /  
0001A
(Routine)
Route 
I91 /  
VT10A
I 00091 ML over I 91 OVER VT 10A

Copyright © 2024 Microsoft and its suppliers. All rights reserved.
Location: I 91 EXIT 13
43.70567, -72.30625
Tea


### Try fitz

In [1496]:
import fitz  # PyMuPDF


document = fitz.open(pdf_path)
    
    # Variable to store the text
text3 = ""
    
    # Iterate through each page
for page_num in range(len(document)):
     # Get the page
    page = document.load_page(page_num)
        
        # Extract text from the page
    page_text = page.get_text()
        
        # Add the page text to our text variable
    text3 += page_text
    
    # Close the document
document.close()
    
text3

'District 4, 27 - WINDSOR County\nOwner: 1 - State Highway Agency\nTown: 158 - NORWICH\nMaintenance Responsibility: 1 - State Highway Agency\nTeam Lead: Stephen Piro,  Inspection Date: 04/15/2024\nStructure #0048N /  0001A(Routine)\nRoute I91 /  VT10A\nI 00091 ML over I 91 OVER VT 10A\nCopyright © 2024 Microsoft and its suppliers. All rights reserved.\nLocation: I 91 EXIT 13\n43.70567, -72.30625\nTeam Lead: Stephen Piro,  Inspection Date: 04/15/2024\nStructure #0048N /  0001A(Routine)\nRoute I91 /  VT10A\nI 00091 ML over I 91 OVER VT 10A\nIDENTIFICATION\n(1) State Names\n50 - Vermont\n(8) Structure Number\n200091048N14112\n(5) Inventory Route\n1\n(2) Highway Agency District\n4 - District 4\n(3) County Code\n27 - WINDSOR\n(4) Place Code\n52900\n(6) Features Intersected\nI 91 OVER VT 10A\n(7) Facility Carried\nI 00091 ML\n(9) Location\nI 91 EXIT 13\n(11) Mile Point\n74.814 mi\n(12) Base Highway Network\nYes\n(13) LRS Inventory Rte & Subrte\n0270000091\n(16) Latitude\n43.7056694444444\n(1

In [1497]:
doc3 = nlp(text3)
doc3

District 4, 27 - WINDSOR County
Owner: 1 - State Highway Agency
Town: 158 - NORWICH
Maintenance Responsibility: 1 - State Highway Agency
Team Lead: Stephen Piro,  Inspection Date: 04/15/2024
Structure #0048N /  0001A(Routine)
Route I91 /  VT10A
I 00091 ML over I 91 OVER VT 10A
Copyright © 2024 Microsoft and its suppliers. All rights reserved.
Location: I 91 EXIT 13
43.70567, -72.30625
Team Lead: Stephen Piro,  Inspection Date: 04/15/2024
Structure #0048N /  0001A(Routine)
Route I91 /  VT10A
I 00091 ML over I 91 OVER VT 10A
IDENTIFICATION
(1) State Names
50 - Vermont
(8) Structure Number
200091048N14112
(5) Inventory Route
1
(2) Highway Agency District
4 - District 4
(3) County Code
27 - WINDSOR
(4) Place Code
52900
(6) Features Intersected
I 91 OVER VT 10A
(7) Facility Carried
I 00091 ML
(9) Location
I 91 EXIT 13
(11) Mile Point
74.814 mi
(12) Base Highway Network
Yes
(13) LRS Inventory Rte & Subrte
0270000091
(16) Latitude
43.7056694444444
(17) Longitude
-72.30625
(98) Border Bridge S

### Try another

In [1498]:
from pdfminer.high_level import extract_text

text4 = extract_text(pdf_path)
doc4 = nlp(text4)
doc4

CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

Route I91 /  VT10A

Structure #0048N /  0001A(Routine)

I 00091 ML over I 91 OVER VT 10A

Team Lead: Stephen Piro,  Inspection Date: 04/15/2024

Town: 158 - NORWICH

District 4, 27 - WINDSOR County

Owner: 1 - State Highway Agency

Maintenance Responsibility: 1 - State Highway Agency


Route I91 /  VT10A

Structure #0048N /  0001A(Routine)

I 00091 ML over I 91 OVER VT 10A

Team Lead: Stephen Piro,  Inspection Date: 04/15/2024

Location: I 91 EXIT 13

Copyright © 2024 Microsoft and its suppliers. All rights reserved.

43.70567, -72.30625


Route I91 /  VT10A

Structure #0048N /  0001A(Routine)

I 00091 ML over I 91 OVER VT 10A

Team Lead: Stephen Piro,  Inspection Date: 04/15/2024

IDENTIFICATION

CLASSIFICATION

(1) State Names
(8) Structure Number
(5) Inventory Route
(2) Highway Agency District
(3) County Code
(4) Place Code
(6) Features Intersected
(7) Facility Carried
(9) Location
(11) Mile Point
(12) Base Highway Network
(13) LRS Inventory Rte & Subrte
(16) Latitude
(17) Longitude

### Extract text better

In [1499]:
text_with_tables = ""
tables = []
with pdfplumber.open(pdf_path) as pdf:
    for page in pdf.pages:
        # Extract tables first
        table_per_page = page.extract_tables()
        if(len(table_per_page) > 0):
            tables.append(table_per_page)
        # # Format tables as text
        for table in tables:
            for row in table:
                for a_row in row:
                    # print(a_row)
        #         # Filter None values and join cells with tabs
            #     print(row)
                    row_text = "\t".join([cell[0] if cell else "" for cell in a_row])
                    text_with_tables += row_text + "\n"
                text_with_tables += "\n"  # Extra line between tables
            
        #     # Extract the remaining text
        text = page.extract_text()
        if text:
            text_with_tables += text + "\n\n"
text_with_tables

CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

'Route I91 / VT10A\nStructure #0048N / 0001A(Routine)\nI 00091 ML over I 91 OVER VT 10A\nTeam Lead: Stephen Piro, Inspection Date: 04/15/2024\nTown: 158 - NORWICH\nDistrict 4, 27 - WINDSOR County\nOwner: 1 - State Highway Agency\nMaintenance Responsibility: 1 - State Highway Agency\n\nRoute I91 / VT10A\nStructure #0048N / 0001A(Routine)\nI 00091 ML over I 91 OVER VT 10A\nTeam Lead: Stephen Piro, Inspection Date: 04/15/2024\nLocation: I 91 EXIT 13\nCopyright © 2024 Microsoft and its suppliers. All rights reserved.\n43.70567, -72.30625\n\nI\n(\n(\n(\nA\nB\nC\n*\n\nRoute I91 / VT10A\nStructure #0048N / 0001A(Routine)\nI 00091 ML over I 91 OVER VT 10A\nTeam Lead: Stephen Piro, Inspection Date: 04/15/2024\nIDENTIFICATION CLASSIFICATION\n(1) State Names 50 - Vermont (112) NBIS Bridge Length Y\n(8) Structure Number 200091048N14112 (104) Highway System 1\n(5) Inventory Route 1 (26) Functional Class 1 - Rural Principal Arterial -\n(2) Highway Agency District 4 - District 4 (100) Defense Highway

In [1500]:
print(tables)
len(tables)

[[[['INSPECTIONS *'], ['(90) Inspection Date 04/15/2024'], ['(91) Frequency 12'], ['(92) Critical Feature Inspection Done Freq. (Mon) Date'], ['A: Fracture Critical Detail No'], ['B: Underwater Inspection No'], ['C: Other Special Inspection'], ['* The inspection date and frequency information in this box contains\nthe current NBI date and frequency information. Please refer to the\nreport header for the date this inspection was conducted.']]], [[[None, None, None, None, None, '', '', ''], ['ELEMENTS', 'DESCRIPTION', 'UNITS', 'TOTAL', 'CS1', 'CS2', 'CS3', 'CS4'], ['12 Reinforced Concrete Deck SF 9450 8730 445 275 0\n1080 Delamination/Spall/Patched Area SF 125 0 0 125 0\n1120 Efflorescence/Rust Staining SF 420 0 320 100 0\n1130 Cracking (RC and Other) SF 175 0 125 50 0\n510 Wearing Surfaces SF 8393 8368 0 25 0\n3210 Delam/Spall/Patched Area/Pothole SF 25 0 0 25 0', None, None, None, None, None, None, None], ['301 Pourable Joint Seal LF 41 9 20 12 0\n2340 Seal Cracking LF 32 0 20 12 0', N

7

### Clean Text

`re.sub()` this stands for subsititue 

In [1501]:
text = re.sub(r'\s+', ' ', text_with_tables)  # Normalize whitespace
text = re.sub(r'(\d) - ', r'\1 - ', text_with_tables)
print(text[:400])

Route I91 / VT10A
Structure #0048N / 0001A(Routine)
I 00091 ML over I 91 OVER VT 10A
Team Lead: Stephen Piro, Inspection Date: 04/15/2024
Town: 158 - NORWICH
District 4, 27 - WINDSOR County
Owner: 1 - State Highway Agency
Maintenance Responsibility: 1 - State Highway Agency

Route I91 / VT10A
Structure #0048N / 0001A(Routine)
I 00091 ML over I 91 OVER VT 10A
Team Lead: Stephen Piro, Inspection Dat


In [1502]:
doc = nlp(text)
# doc

In [1503]:
text2 = re.sub(r'\s+', ' ', text2)  # Normalize whitespace
text2 = re.sub(r'(\d) - ', r'\1 - ', text2)
print(text2[:400])

District 4, 27 - WINDSOR County Owner: 1 - State Highway Agency Town: 158 - NORWICH Maintenance Responsibility: 1 - State Highway Agency Team Lead: Stephen Piro, Inspection Date: 04/15/2024 Structure #0048N / 0001A (Routine) Route I91 / VT10A I 00091 ML over I 91 OVER VT 10A Copyright © 2024 Microsoft and its suppliers. All rights reserved. Location: I 91 EXIT 13 43.70567, -72.30625 Team Lead: Ste


In [1504]:
doc2 = nlp(text2)
# doc2

In [1505]:
doc

Route I91 / VT10A
Structure #0048N / 0001A(Routine)
I 00091 ML over I 91 OVER VT 10A
Team Lead: Stephen Piro, Inspection Date: 04/15/2024
Town: 158 - NORWICH
District 4, 27 - WINDSOR County
Owner: 1 - State Highway Agency
Maintenance Responsibility: 1 - State Highway Agency

Route I91 / VT10A
Structure #0048N / 0001A(Routine)
I 00091 ML over I 91 OVER VT 10A
Team Lead: Stephen Piro, Inspection Date: 04/15/2024
Location: I 91 EXIT 13
Copyright © 2024 Microsoft and its suppliers. All rights reserved.
43.70567, -72.30625

I
(
(
(
A
B
C
*

Route I91 / VT10A
Structure #0048N / 0001A(Routine)
I 00091 ML over I 91 OVER VT 10A
Team Lead: Stephen Piro, Inspection Date: 04/15/2024
IDENTIFICATION CLASSIFICATION
(1) State Names 50 - Vermont (112) NBIS Bridge Length Y
(8) Structure Number 200091048N14112 (104) Highway System 1
(5) Inventory Route 1 (26) Functional Class 1 - Rural Principal Arterial -
(2) Highway Agency District 4 - District 4 (100) Defense Highway 1 - The inventory route is on
(3) 

In [1506]:
pattern = f"Structure\s+#(\d+)"
matches = re.finditer(pattern, doc.text, re.IGNORECASE)
structure_numbers = []
for match in matches:
    if match.group(1) not in structure_numbers:
        structure_numbers.append(match.group(1))
bridge_id = structure_numbers[0] if len(structure_numbers) == 1 else "Unknown"
bridge_id

<>:1: SyntaxWarning: invalid escape sequence '\s'
<>:1: SyntaxWarning: invalid escape sequence '\s'
C:\Users\MK\AppData\Local\Temp\ipykernel_3172\608127002.py:1: SyntaxWarning: invalid escape sequence '\s'
  pattern = f"Structure\s+#(\d+)"


'0048'

## Find section boundaries

In [1507]:
sentences = list(doc.sents) # this is how you get the sentences
sentences[0].start_char

0

In [1508]:
# senteences = list(doc.sents)
lines = doc.text.split('\n')
section_boundaries = []

section_headers = [
        "deck", "approach", "superstructure", "substructure", 
        "channel", "general observation","channel profile",
    ]
for i, line in enumerate(lines):
    clean_line = line.strip().lower()
    if not clean_line:
        continue
    for header in section_headers:
        if clean_line == header:
            section_boundaries.append((i, header.upper()))
            break
section_boundaries.sort(key=lambda x: x[0])
print(section_boundaries)

[(124, 'DECK'), (199, 'APPROACH'), (238, 'SUPERSTRUCTURE'), (299, 'SUBSTRUCTURE'), (332, 'CHANNEL'), (337, 'GENERAL OBSERVATION'), (574, 'CHANNEL PROFILE')]


In [1509]:
section= {}
sections = []

for i, (line_idx, header) in enumerate(section_boundaries):
    end_idx = section_boundaries[i + 1][0] if  i < len(section_boundaries) - 1 else len(lines)
    
    section = {
        "index": (line_idx, end_idx),
        "content_array": lines[line_idx:end_idx],
        "content": "\n".join(lines[line_idx:end_idx]),
        "header": header
    }
    sections.append(section)

In [1510]:
print(sections[idx_section]["content"])

Substructure
ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4
205 Reinforced Concrete Column EA 6 6 0 0 0
215 Reinforced Concrete Abutment LF 92 0 12 60 20
1080 Delamination/Spall/Patched Area LF 72 0 12 40 20
1130 Cracking (RC and Other) LF 20 0 0 20 0
234 Reinforced Concrete Pier Cap LF 92 78 8 6 0
1130 Cracking (RC and Other) LF 14 0 8 6 0
800 Reinforced Concrete Wing/Retaining Wall EA 4 0 0 4 0
1080 Delamination/Spall/Patched Area EA 4 0 0 4 0
60 - Substructure (4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.)
Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to
heavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat
have heavy spalling that has exposed the steel reinforcing below.
Reinforced concrete abutment #2 is in poor condition having heavy deterioration and decay present scattered throughout.
Abutment stem has heavy

## Get table and description boundary

we will use units here because these are present for everyt row 

In [1511]:
units_patterns = [
        r'\bLF\b',    # Linear Feet
        r'\bEA\b',    # Each
        r'\bSF\b',    # Square Feet
        r'\bCF\b',    # Cubic Feet
        r'\bFT\b',    # Feet 
        r'\bIN\b',    # Inches
        r'\bM\b',     # Meters
        r'\bMM\b'     # Millimeters
    ]
units_regex = '|'.join(units_patterns)

In [1512]:
units_regex

'\\bLF\\b|\\bEA\\b|\\bSF\\b|\\bCF\\b|\\bFT\\b|\\bIN\\b|\\bM\\b|\\bMM\\b'

A simple check to asses if content has a table or not

In [1513]:
if "LF" in sections[idx_section]["content"]:
    print("Yeah")
else:
    print("Nah")

Yeah


### Get last table line

In [1514]:
doc_ = nlp(sections[idx_section]["content"])
lines = doc_.text.split("\n")

last_table_line = -1
for idx, line in enumerate(lines):
    # print(re.search(units_regex, line))
    if re.search(units_regex, line):
        print(line)
        last_table_line = idx

205 Reinforced Concrete Column EA 6 6 0 0 0
215 Reinforced Concrete Abutment LF 92 0 12 60 20
1080 Delamination/Spall/Patched Area LF 72 0 12 40 20
1130 Cracking (RC and Other) LF 20 0 0 20 0
234 Reinforced Concrete Pier Cap LF 92 78 8 6 0
1130 Cracking (RC and Other) LF 14 0 8 6 0
800 Reinforced Concrete Wing/Retaining Wall EA 4 0 0 4 0
1080 Delamination/Spall/Patched Area EA 4 0 0 4 0


In [1515]:
lines

['Substructure',
 'ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4',
 '205 Reinforced Concrete Column EA 6 6 0 0 0',
 '215 Reinforced Concrete Abutment LF 92 0 12 60 20',
 '1080 Delamination/Spall/Patched Area LF 72 0 12 40 20',
 '1130 Cracking (RC and Other) LF 20 0 0 20 0',
 '234 Reinforced Concrete Pier Cap LF 92 78 8 6 0',
 '1130 Cracking (RC and Other) LF 14 0 8 6 0',
 '800 Reinforced Concrete Wing/Retaining Wall EA 4 0 0 4 0',
 '1080 Delamination/Spall/Patched Area EA 4 0 0 4 0',
 '60 - Substructure (4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.)',
 'Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to',
 'heavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat',
 'have heavy spalling that has exposed the steel reinforcing below.',
 'Reinforced concrete abutment #2 is in poor condition having heavy deterioration and d

In [1516]:
lines[last_table_line]

'1080 Delamination/Spall/Patched Area EA 4 0 0 4 0'

In [1517]:

table = lines[:last_table_line+1]
desc = lines[last_table_line+1:]
table, desc

(['Substructure',
  'ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4',
  '205 Reinforced Concrete Column EA 6 6 0 0 0',
  '215 Reinforced Concrete Abutment LF 92 0 12 60 20',
  '1080 Delamination/Spall/Patched Area LF 72 0 12 40 20',
  '1130 Cracking (RC and Other) LF 20 0 0 20 0',
  '234 Reinforced Concrete Pier Cap LF 92 78 8 6 0',
  '1130 Cracking (RC and Other) LF 14 0 8 6 0',
  '800 Reinforced Concrete Wing/Retaining Wall EA 4 0 0 4 0',
  '1080 Delamination/Spall/Patched Area EA 4 0 0 4 0'],
 ['60 - Substructure (4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.)',
  'Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to',
  'heavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat',
  'have heavy spalling that has exposed the steel reinforcing below.',
  'Reinforced concrete abutment #2 is in poor condition having heavy det

In [1518]:
table_text = '\n'.join(table)
print(table_text)

Substructure
ELEMENTS DESCRIPTION UNITS TOTAL CS1 CS2 CS3 CS4
205 Reinforced Concrete Column EA 6 6 0 0 0
215 Reinforced Concrete Abutment LF 92 0 12 60 20
1080 Delamination/Spall/Patched Area LF 72 0 12 40 20
1130 Cracking (RC and Other) LF 20 0 0 20 0
234 Reinforced Concrete Pier Cap LF 92 78 8 6 0
1130 Cracking (RC and Other) LF 14 0 8 6 0
800 Reinforced Concrete Wing/Retaining Wall EA 4 0 0 4 0
1080 Delamination/Spall/Patched Area EA 4 0 0 4 0


In [1519]:
desc_text = '\n'.join(desc)
print(desc_text)

60 - Substructure (4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.)
Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to
heavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat
have heavy spalling that has exposed the steel reinforcing below.
Reinforced concrete abutment #2 is in poor condition having heavy deterioration and decay present scattered throughout.
Abutment stem has heavy spalling along the outer portions that have exposed the steel reinforcing with measurable
section loss. Bridge seat has heavy spalling along the outer portions that have undermined bearing #1 nearly 80% and
bearing #6 nearly 70% and bearing #4 has ~50% undermining due to heavy spalling and pooling water along the bridge
seat.
A71 - Abutment End Walls Condition (Poor)
Reinforced concrete backwalls are in poor condition having heavy concrete spallin

## Get bridge components from tables

### Create component regex

**Important** - double check this code here you changed the order of bridge components

In [1520]:
processed_components =[]
bridge_components = [
    "Deck",
    "Open Girder/Beam",
    "Girder/Beam",
    "Assembly Joint without Seal",
    "Pourable Joint Seal",
    "Open Expansion Joint",
    "Joint",
    "Truss",
    "Column",
    "Arch",
    "Pier Cap",
    "Pier Wall",
    "Pier",
    "Abutment",
    "Bearing",
    "Pile",
    "Railing",
    "Pile Cap",
    "Bent",
    "Parapet/Guard Rail",
    "Expansion Joint",
    "Foundation",
    "Backwall",
    "Diaphragm",
    "Fascia",
    "Elastomeric Bearing",
    "Approach",
    "Slab",
    "Approach Slab",
    "Wingwall",
    "Spandrel",
    "Cantilever",
    "Movable Bearing",
    "Fixed Bearing",
    "Counterweight",
    "Cutwater",
    "Flood Arches",
    "Suspension Cable",
    "Tower",
    "Cable-Stay",
    "Wing/Retaining Wall",
    "Bearing Pedestal",
    "Stringer",
    "Footing",
    "Shear Key",
    "Drainage System",
    "Seismic Device",
    "Crash Barrier",
    "Culvert",
    "Culvert Pipe",
    "Headwall",
    "Apron",
    "Invert",
    "Soffit",
    "Skewback",
    "Keystone",
    "Camber",
    "Tie Rod",
    "Dampener"
]
bridge_components = sorted(list(set(bridge_components)), key=len, reverse=True)
materials = ["Reinforced Concrete", "Prestressed Concrete", "RC", "PC", "Steel", "Concrete", "Timber", "Wood", "Metal", "Metal Bridge"]


In [1521]:
for component in bridge_components:
        # Handle slash-separated components
        if "/" in component:
            parts = [re.escape(part.strip()) for part in component.split("/")]
            # Create pattern that allows for typos or partial matches
            pattern = f"(?:{'|'.join(parts)}|{re.escape(component)})"
            
            # Special case for "Wing/Retaining Wall" - also match if "Wall" has a typo
            if "Wall" in component:
                wall_variations = ["Wall", "Wal", "Walls"]
                wall_pattern = f"(?:{'|'.join(wall_variations)})"
                # Replace "Wall" in the pattern with the variations
                pattern = pattern.replace(re.escape("Wall"), wall_pattern)
        else:
            pattern = re.escape(component)
        
        # Make trailing 's' optional for plurals (e.g. match "Piers" with "Pier")
        if pattern.endswith("s\\"):
            pattern = pattern[:-2] + "s?"
        else:
            pattern = pattern + "s?"
            
        # Remove number if present (e.g. "Abutment 1" -> "Abutment( \d+)?")
        if " " in pattern and any(c.isdigit() for c in pattern):
            pattern = re.sub(r'\\\ \d+', '', pattern) + r'(\ \d+)?'
            
        processed_components.append(pattern)

In [1522]:
material_pattern = f"(?:{'|'.join(materials)})\\s+"
component_pattern = f"(?:{'|'.join(processed_components)})"
    
# Final pattern including optional material prefix and component
final_pattern = fr"(\d+\s+(?:{material_pattern})?{component_pattern})"

In [1523]:
component_sections = []
doc_t = nlp(table_text)
    
matches = list(re.finditer(final_pattern, doc_t.text))
for i, match in enumerate(matches):
    start_pos = match.start()
        
    # Determine end position (either next component or end of text)
    if i < len(matches) - 1:
        end_pos = matches[i+1].start()
    else:
        end_pos = len(table_text)
        
    # Extract the full section text
    section_text = table_text[start_pos:end_pos].strip()
        
    # Extract just the component name (may include material)
    component_text = match.group(1)
        
    component_sections.append({
            "full_section": section_text,
            "component": component_text,
            "start": start_pos,
            "end": end_pos
        })

put in a check for length of components has to nbe more than one to continue

In [1524]:
# print(component_sections[0]["full_section"])

In [1525]:
# print(component_sections[1]["full_section"])

## Get Component and defect Info

### Component info

In [1526]:
# Regex patterns for extracting data
id_pattern = r"^(\d+)\s+"
units_pattern = r"\b(LF|SF|EA|CF|FT|IN)\b"
values_pattern = r"\b(\d+(?:\s+\d+)*)\s*$"

In [1527]:
def extract_component_values(component_text, full_section, component_data):
    """
    Extract component values directly after the component name and unit.
    
    Args:
        component_text (str): The component text (e.g. "215 Reinforced Concrete Abutment")
        full_section (str): The full text of the component section
        component_data (dict): The component data dictionary to update
    """
    # Escape special characters in the component text for regex
    escaped_component = re.escape(component_text)
    
    # Pattern to find the component's unit and values
    # Look for exactly 5 numbers after the unit (TOTAL + CS1-CS4)
    pattern = fr"{escaped_component}\s+(LF|SF|EA|CF|FT|IN)\s+(\d+\s+\d+\s+\d+\s+\d+\s+\d+)"
    match = re.search(pattern, full_section)
    
    if match:
        component_data["units"] = match.group(1)
        values_str = match.group(2)
        component_data["values"] = [int(val) for val in values_str.split()]
    else:
        # Fallback pattern for cases with fewer or more values
        pattern = fr"{escaped_component}\s+(LF|SF|EA|CF|FT|IN)\s+(\d+(?:\s+\d+)*?)(?:\s+\d{{3,4}}\s|$)"
        match = re.search(pattern, full_section)
        
        if match:
            component_data["units"] = match.group(1)
            values_str = match.group(2)
            component_data["values"] = [int(val) for val in values_str.split()]
        else:
            # Last resort fallback
            unit_pattern = r"\b(LF|SF|EA|CF|FT|IN)\s+(\d+(?:\s+\d+){0,4})"
            pos = full_section.find(component_text) + len(component_text)
            remaining_text = full_section[pos:pos+30]  # Look at limited text to avoid capturing too much
            unit_match = re.search(unit_pattern, remaining_text)
            
            if unit_match:
                component_data["units"] = unit_match.group(1)
                values_str = unit_match.group(2)

                component_data["values"] = [int(val) for val in values_str.split()]

**Important** -- double check this you changed some stuff here

In [1528]:
def extract_defects(section_text, component_data):
    """
    Extract defect information from a component section.
    
    Args:
        section_text (str): The full component section text
        component_data (dict): The component data dictionary to update
    """
    # Pattern to match defect entries: ID + Description + Units + Values
    # Use a more specific pattern to match exactly the format we expect
    defect_pattern = r"(\d{3,4})\s+([A-Za-z][A-Za-z\s\/\-\(\)]+)\s+(LF|SF|EA|CF|FT|IN)\s+(\d+(?:\s+\d+){0,4})" # new pattern
    # defect_pattern = r"(\d{3,4})\s+([^L][^F][^S][^E][^A][^\d]+?)\s+(LF|SF|EA|CF|FT|IN)\s+(\d+(?:\s+\d+){0,4})" # old pattern
    
    # Find all defects in the section text
    defect_matches = re.finditer(defect_pattern, section_text)
    
    for match in defect_matches:
        defect_id = match.group(1)
        defect_desc = match.group(2).strip()
        defect_unit = match.group(3)
        defect_values_str = match.group(4)
        defect_values = [int(val) for val in defect_values_str.split()]
        
        # Skip if this is the component ID itself (first entry)
        if defect_id == component_data.get("id"):
            continue
            
        # Create defect entry
        defect = {
            "id": defect_id,
            "description": defect_desc,
            "unit": defect_unit,
            "total": defect_values[0] if defect_values else None,
            "cs_values": defect_values[1:] if len(defect_values) > 1 else []
        }
        
        component_data["defects"].append(defect)

In [1529]:
component_sections

[{'full_section': '205 Reinforced Concrete Column EA 6 6 0 0 0',
  'component': '205 Reinforced Concrete Column',
  'start': 62,
  'end': 106},
 {'full_section': '215 Reinforced Concrete Abutment LF 92 0 12 60 20\n1080 Delamination/Spall/Patched Area LF 72 0 12 40 20\n1130 Cracking (RC and Other) LF 20 0 0 20 0',
  'component': '215 Reinforced Concrete Abutment',
  'start': 106,
  'end': 254},
 {'full_section': '234 Reinforced Concrete Pier Cap LF 92 78 8 6 0\n1130 Cracking (RC and Other) LF 14 0 8 6 0',
  'component': '234 Reinforced Concrete Pier Cap',
  'start': 254,
  'end': 345},
 {'full_section': '800 Reinforced Concrete Wing/Retaining Wall EA 4 0 0 4 0\n1080 Delamination/Spall/Patched Area EA 4 0 0 4 0',
  'component': '800 Reinforced Concrete Wing',
  'start': 345,
  'end': 451}]

In [1530]:
structured_components = []
for section in component_sections:
    print( section["component"])
    component_data = {
            "text": section["component"],
            "full_section": section["full_section"],
            "defects": []  # Will hold defect information
        }
        
    # Extract ID
    id_match = re.search(id_pattern, section["component"])
    if id_match:
        print(id_match.group(1))
        component_data["id"] = id_match.group(1)
    units_match = re.search(units_pattern, section["full_section"])
    if units_match:
        print(units_match.group(1))
        component_data["units"] = units_match.group(1)
    values_match = re.search(values_pattern, section["full_section"])
    extract_component_values(section["component"], section["full_section"], component_data)
    extract_defects(section["full_section"], component_data)
    structured_components.append(component_data)

205 Reinforced Concrete Column
205
EA
215 Reinforced Concrete Abutment
215
LF
234 Reinforced Concrete Pier Cap
234
LF
800 Reinforced Concrete Wing
800
EA


Expetected Output
`[{'text': '215 Reinforced Concrete Abutment',
  'full_section': '215 Reinforced Concrete Abutment LF 51 42 6 3 0\n1080 Delamination/Spall/Patched Area LF 3 0 0 3 0\n1090 Exposed Rebar LF 1 0 1 0 0\n1130 Cracking (RC and Other) LF 5 0 5 0 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 3,
    'cs_values': [0, 0, 3, 0]},
   {'id': '1090',
    'description': 'Exposed Rebar',
    'unit': 'LF',
    'total': 1,
    'cs_values': [0, 1, 0, 0]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'LF',
    'total': 5,
    'cs_values': [0, 5, 0, 0]}],
  'id': '215',
  'units': 'LF',
  'values': [51, 42, 6, 3, 0]},
 {'text': '800 Reinforced Concrete Wing',
  'full_section': '800 Reinforced Concrete Wing/Retaining Wall EA 4 0 4 0 0\n1120 Efflorescence/Rust Staining EA 2 0 2 0 0\n1130 Cracking (RC and Other) EA 2 0 2 0 0',
  'defects': [{'id': '1120',
    'description': 'Efflorescence/Rust Staining',
    'unit': 'EA',
    'total': 2,
    'cs_values': [0, 2, 0, 0]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'EA',
    'total': 2,
    'cs_values': [0, 2, 0, 0]}],
  'id': '800',
  'units': 'EA',
  'values': [4, 0, 4, 0, 0]}]`

In [1531]:
structured_components

[{'text': '205 Reinforced Concrete Column',
  'full_section': '205 Reinforced Concrete Column EA 6 6 0 0 0',
  'defects': [],
  'id': '205',
  'units': 'EA',
  'values': [6, 6, 0, 0, 0]},
 {'text': '215 Reinforced Concrete Abutment',
  'full_section': '215 Reinforced Concrete Abutment LF 92 0 12 60 20\n1080 Delamination/Spall/Patched Area LF 72 0 12 40 20\n1130 Cracking (RC and Other) LF 20 0 0 20 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 72,
    'cs_values': [0, 12, 40, 20]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'LF',
    'total': 20,
    'cs_values': [0, 0, 20, 0]}],
  'id': '215',
  'units': 'LF',
  'values': [92, 0, 12, 60, 20]},
 {'text': '234 Reinforced Concrete Pier Cap',
  'full_section': '234 Reinforced Concrete Pier Cap LF 92 78 8 6 0\n1130 Cracking (RC and Other) LF 14 0 8 6 0',
  'defects': [{'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit

## Get info from description text

### get contexts

In [1532]:
desc_text.split("\n")

['60 - Substructure (4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.)',
 'Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to',
 'heavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat',
 'have heavy spalling that has exposed the steel reinforcing below.',
 'Reinforced concrete abutment #2 is in poor condition having heavy deterioration and decay present scattered throughout.',
 'Abutment stem has heavy spalling along the outer portions that have exposed the steel reinforcing with measurable',
 'section loss. Bridge seat has heavy spalling along the outer portions that have undermined bearing #1 nearly 80% and',
 'bearing #6 nearly 70% and bearing #4 has ~50% undermining due to heavy spalling and pooling water along the bridge',
 'seat.',
 'A71 - Abutment End Walls Condition (Poor)',
 'Reinforced concrete backwalls are in po

**Important** - you added this ti make sure there is not much overflow double check this change

Find a better way to do this 

In [1533]:
def get_team_index(lines):
    for idx, line in enumerate(lines):
        if line.startswith('Team Lead:'):
            return idx
    return None

In [1534]:
def check_for_team(lines):
    for line  in lines:
        if line.startswith('Team Lead:'):
            return True
    return False

In [1535]:
if '' in desc_text.split("\n") and check_for_team(desc_text.split("\n")):
   print("yeah")
else:
   print("nah")

nah


In [1536]:
desc_text.split("\n")

['60 - Substructure (4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.)',
 'Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to',
 'heavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat',
 'have heavy spalling that has exposed the steel reinforcing below.',
 'Reinforced concrete abutment #2 is in poor condition having heavy deterioration and decay present scattered throughout.',
 'Abutment stem has heavy spalling along the outer portions that have exposed the steel reinforcing with measurable',
 'section loss. Bridge seat has heavy spalling along the outer portions that have undermined bearing #1 nearly 80% and',
 'bearing #6 nearly 70% and bearing #4 has ~50% undermining due to heavy spalling and pooling water along the bridge',
 'seat.',
 'A71 - Abutment End Walls Condition (Poor)',
 'Reinforced concrete backwalls are in po

In [1537]:
if '' in desc_text.split("\n") and check_for_team(desc_text.split("\n")):
    clean_idx_start = desc_text.split("\n").index('')
    lines = desc_text.split("\n")
    clean_idx_end = get_team_index(lines)
    desc_text_start = desc_text.split("\n")[:clean_idx_start]
    if len(lines) > clean_idx_end:
        desc_text_end = desc_text.split("\n")[clean_idx_end+1:]
        desc_test = desc_text_start + desc_text_end
    else:
        desc_test = desc_text_start
elif '' in desc_text.split("\n"):
    cleaned_idx = desc_text.split("\n").index('')
    desc_test = desc_text.split("\n")[:cleaned_idx]
else:
    cleaned_idx = len(desc_text.split("\n")) + 1
    desc_test = desc_text.split("\n")[:cleaned_idx]
desc_text = "\n".join(desc_test)

In [1538]:
print(desc_text)

60 - Substructure (4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.)
Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to
heavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat
have heavy spalling that has exposed the steel reinforcing below.
Reinforced concrete abutment #2 is in poor condition having heavy deterioration and decay present scattered throughout.
Abutment stem has heavy spalling along the outer portions that have exposed the steel reinforcing with measurable
section loss. Bridge seat has heavy spalling along the outer portions that have undermined bearing #1 nearly 80% and
bearing #6 nearly 70% and bearing #4 has ~50% undermining due to heavy spalling and pooling water along the bridge
seat.
A71 - Abutment End Walls Condition (Poor)
Reinforced concrete backwalls are in poor condition having heavy concrete spallin

In [1539]:
def parse_condition(condition_text):
    """
    Parse condition text to extract rating and assessment description.
    
    Args:
        condition_text (str): The condition text from the header
        
    Returns:
        tuple: (condition_rating, assessment_description)
    """
    if not condition_text:
        return None, None
    
    # Handle "N/A" or "Not Applicable" cases
    if condition_text.upper() in ["N/A", "NOT APPLICABLE"]:
        return "Not Applicable", None
    
    # Handle simple condition ratings without explanations
    if condition_text.upper() in ["SATISFACTORY", "GOOD", "FAIR", "POOR"]:
        return condition_text.upper(), None
    
    # Pattern for conditions with numeric codes: "6 - SATISFACTORY CONDITION - explanation"
    code_pattern = r'(?:(\d+)\s*-\s*)?([A-Z\s]+)(?:\s*-\s*(.+))?'
    match = re.match(code_pattern, condition_text)
    
    if match:
        code = match.group(1)  # Numeric code like "6"
        rating = match.group(2).strip()  # Rating like "SATISFACTORY CONDITION"
        assessment = match.group(3)  # Explanation text
        
        # Normalize rating text (remove "CONDITION" if present)
        rating = rating.replace("CONDITION", "").strip()
        
        return rating, assessment
    
    # Default case - return the whole text as the rating
    return condition_text, None

In [1540]:
# header_pattern = r'([A-Z]?\d+\s+-\s+[^()]+)(?:\s*\(([^)]+)\))?'
    
#     # Find all headers in the text
# header_matches = list(re.finditer(header_pattern, desc_text))
    
#     # If no headers found, return the whole text as one section
# if not header_matches:
#     desc_sections =  [{"header": None, "condition": None, "content": text.strip()}]
    
#     # Create sections by finding the text between headers
# desc_sections = []
    
# for i, match in enumerate(header_matches):
#     header = match.group(1).strip()
#     condition_text = match.group(2).strip() if match.group(2) else None
        
#         # Parse the condition and assessment
#     condition_rating, assessment = parse_condition(condition_text)
        
#         # Find the start of the content (after the header and condition)
#     content_start = match.end()
        
#         # Find the end of the content (start of next header or end of text)
#     if i < len(header_matches) - 1:
#         content_end = header_matches[i + 1].start()
#     else:
#         content_end = len(text)
        
#         # Extract the content
#     content = desc_text[content_start:content_end].strip()
        
#     desc_sections.append({
#             "header": header,
#             "condition_text": condition_text,
#             "condition_rating": condition_rating,
#             "assessment": assessment,
#             "content": content
#         })
# desc_sections

**Important** --- Retry this code make sure it works with substructure

In [1541]:
header_pattern = r'(?:([A-Z](?:\.[A-Z])?(?:\.\d+)?(?:\s+-)?|\d+\s+-)\s+([^()]+))(?:\s*\(([^)]+)\))?'

header_matches = list(re.finditer(header_pattern, desc_text))
    
    # If no headers found, return the whole text as one section
if not header_matches:
    desc_sections = [{"header": None, "condition_text": None, "condition_rating": None, 
                "assessment": None, "content": text.strip()}]
    
    # Create sections by finding the text between headers
desc_sections = []
    
for i, match in enumerate(header_matches):
        # Combine the prefix and title parts to form the complete header
    prefix = match.group(1).strip() if match.group(1) else ""
    title = match.group(2).strip() if match.group(2) else ""
    header = f"{prefix} {title}".strip()
        
    condition_text = match.group(3).strip() if match.group(3) else None
        
        # Parse the condition and assessment
    condition_rating, assessment = parse_condition(condition_text)
        
        # Find the start of the content (after the header and condition)
    content_start = match.end()
        
        # Find the end of the content (start of next header or end of text)
    if i < len(header_matches) - 1:
        content_end = header_matches[i + 1].start()
    else:
        content_end = len(text)
        
        # Extract the content
    content = desc_text[content_start:content_end].strip()
        
    desc_sections.append({
            "header": header,
            "condition_text": condition_text,
            "condition_rating": condition_rating,
            "assessment": assessment,
            "content": content if len(content) > 5 else ""
        })
    
desc_sections

[{'header': '60 - Substructure',
  'condition_text': '4 - POOR CONDITION - advanced section loss, deterioration, spalling or scour.',
  'condition_rating': 'POOR',
  'assessment': 'advanced section loss, deterioration, spalling or scour.',
  'content': 'Reinforced concrete abutment #1 is in fair condition having heavy deterioration present scattered throughout with minor to\nheavy cracking, rust staining and efflorescence leakage along the stem. Outer portions of abutment stem / bridge seat\nhave heavy spalling that has exposed the steel reinforcing below.\nReinforced concrete abutment #2 is in poor condition having heavy deterioration and decay present scattered throughout.\nAbutment stem has heavy spalling along the outer portions that have exposed the steel reinforcing with measurable\nsection loss. Bridge seat has heavy spalling along the outer portions that have undermined bearing #1 nearly 80% and\nbearing #6 nearly 70% and bearing #4 has ~50% undermining due to heavy spalling an

## Matching components and desciption

### Extract component name

remeber here we would neeed to extend to specoal cases

In [1542]:
def extract_component_name(text):
    """
    Extract the core component name from the text, removing ID and material prefixes.
    
    Args:
        text (str): The component text (e.g., "215 Reinforced Concrete Abutment")
        
    Returns:
        str: The core component name (e.g., "Abutment")
    """
    # Remove leading ID number
    text = re.sub(r'^\d+\s+', '', text)
    
    # Remove common material prefixes
    for material in materials:
        text = text.replace(material, '').strip()

    # we will need to extend special cases
    if text == "Open Girder" or text == "Girder":
        text = "Beam"
    if "Bearing" in text:
        text = "Bearing"
    
    # Handle special case for "Wing/Retaining Wall"
    if "Wing" in text and "Wall" not in text:
        text = "Wing/Retaining Wall"
    
    return text

In [1543]:
len(structured_components)

4

In [1544]:
# print(structured_components[2]['text'])
# extract_component_name(structured_components[2]['text'])

### Generate component variations

what about lowercase with s and no s?


Rememer here we would dstill need to special cases

In [1545]:
def generate_component_variations(component_name):
    """
    Generate variations of a component name for flexible matching.
    
    Args:
        component_name (str): The core component name
        
    Returns:
        list: List of possible variations of the component name
    """
    variations = [component_name]

    if component_name == "Beam":
        component_names = ["Beam", "Girder"]

        for c_name in component_names:
            variations.append(c_name.lower())
    
        # Handle singular/plural variations
        if c_name.endswith('s'):
            variations.append(c_name[:-1])  # Remove 's'
        else:
            variations.append(c_name + 's')  # Add 's'
            variations.append(c_name + r' \d+')
        variations.append(c_name + r' \d+') 
    else:
    
        # Handle variations with numbers
        variations.append(component_name + r' \d+')  # Component followed by number
        
        # Add lowercase version
        variations.append(component_name.lower())
        
        # Handle singular/plural variations
        if component_name.endswith('s'):
            variations.append(component_name[:-1])  # Remove 's'
        else:
            variations.append(component_name + 's')  # Add 's'
        
        # Handle variations with numbers
        variations.append(component_name + r' \d+')  # Component followed by number
    
    # Handle compound components with slashes
    if '/' in component_name:
        parts = [part.strip() for part in component_name.split('/')]
        variations.extend(parts)
        
        # Special case for "Wing/Retaining Wall"
        if "Wing" in component_name and "Wall" in component_name:
            variations.append("Wingwall")
    return variations

In [1546]:
# print(extract_component_name(structured_components[0]['text']))
# generate_component_variations(extract_component_name(structured_components[0]['text']))

### Match comoponent name in headers

**Important** -- you changed code here make sure it works for older pdfs

In [1547]:
def find_in_headers(component_variations, description_sections):
    """
    Try to find component variations in the section headers.
    
    Args:
        component_variations (list): List of component name variations
        description_sections (list): List of section dictionaries
        
    Returns:
        dict or None: The matching section or None if no match found
    """
    for section in description_sections:
        header = section["header"]
        header_2 = ''
        trough_joint = False
        
        if not header:  # to check if a string exists in a larger string just do this 
            continue
        # print(header)
        if "joint" in header.lower() and "trough" in header.lower():
            # print("here")
            # print(header)    
            header = "Open Expansion Joint"
            header_2 = "Assembly Joint without Seal"
            trough_joint = True
        elif "joint" in header.lower() and "bridge" in header.lower():
            header = "Pourable Joint Seal"
        # Check if any variation appears in the header
        for variation in component_variations:
            # print(variation)
            if trough_joint:
                if variation.lower() in header.lower() or variation.lower() in header_2.lower():
                    # print("here 2")
                    return section
            else:
                if variation.lower() in header.lower():
                    # print("here 2")
                    return section
    
    return None

In [1548]:
def check_comp_desc_matching(idx):
    print(structured_components[idx]['text'])
    return find_in_headers(generate_component_variations(extract_component_name(structured_components[idx]['text'])), desc_sections)

In [1549]:
# check_comp_desc_matching(2)

### Find in content

Double check  what is going here there might be an issue

In [1550]:
def find_in_content(component_variations, description_sections):
    """
    Try to find component variations in the section content.
    
    Args:
        component_variations (list): List of component name variations
        description_sections (list): List of section dictionaries
        
    Returns:
        dict or None: The matching section or None if no match found
    """
    for section in description_sections:
        content = section["content"]
        
        if not content:
            continue
            
        # Check if any variation appears in the content
        for variation in component_variations:
            # Create a pattern that matches the variation as a whole word
            pattern = r'\b' + variation + r'\b'
            if re.search(pattern, content, re.IGNORECASE):
                return section
                
            # For patterns with \d+, we need a special approach
            if r'\d+' in variation:
                base = variation.replace(r'\d+', '')
                pattern = r'\b' + base + r'\d+\b'
                if re.search(pattern, content, re.IGNORECASE):
                    return section
    
    return None

In [1551]:
def check_comp_desc_matching_content(idx):
    print(structured_components[idx]['text'])
    return find_in_content(generate_component_variations(extract_component_name(structured_components[idx]['text'])), desc_sections)

In [1552]:
# check_comp_desc_matching_content(0)

In [1553]:
updated_components = []

for component in structured_components:
        # Create a copy of the component data
        updated_component = component.copy()
        
        # Extract the component name without the ID and material prefixes
        component_name = extract_component_name(component["text"])
        component_variations = generate_component_variations(component_name)
        
        # Try to find a match in the section headers first
        header_match = find_in_headers(component_variations, desc_sections)
        
        if header_match:
            # If found in a header, add the section information
            updated_component["description_section"] = header_match
        else:
            # If not found in headers, try to find mentions in the content
            content_match = find_in_content(component_variations, desc_sections)
            
            if content_match:
                updated_component["description_section"] = content_match
            else:
                updated_component["description_section"] = {}
        
        updated_components.append(updated_component)
updated_components

[{'text': '205 Reinforced Concrete Column',
  'full_section': '205 Reinforced Concrete Column EA 6 6 0 0 0',
  'defects': [],
  'id': '205',
  'units': 'EA',
  'values': [6, 6, 0, 0, 0],
  'description_section': {'header': '85 - Pier Columns Condition',
   'condition_text': 'Good',
   'condition_rating': 'GOOD',
   'assessment': None,
   'content': 'Both piers have three (3) concrete columns that are in fairly good condition.'}},
 {'text': '215 Reinforced Concrete Abutment',
  'full_section': '215 Reinforced Concrete Abutment LF 92 0 12 60 20\n1080 Delamination/Spall/Patched Area LF 72 0 12 40 20\n1130 Cracking (RC and Other) LF 20 0 0 20 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 72,
    'cs_values': [0, 12, 40, 20]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'LF',
    'total': 20,
    'cs_values': [0, 0, 20, 0]}],
  'id': '215',
  'units': 'LF',
  'values': [92, 0, 12, 60, 20]

## Get maintenance reccomendations

In [1554]:
doc4

Route I91 /  VT10A

Structure #0048N /  0001A(Routine)

I 00091 ML over I 91 OVER VT 10A

Team Lead: Stephen Piro,  Inspection Date: 04/15/2024

Town: 158 - NORWICH

District 4, 27 - WINDSOR County

Owner: 1 - State Highway Agency

Maintenance Responsibility: 1 - State Highway Agency


Route I91 /  VT10A

Structure #0048N /  0001A(Routine)

I 00091 ML over I 91 OVER VT 10A

Team Lead: Stephen Piro,  Inspection Date: 04/15/2024

Location: I 91 EXIT 13

Copyright © 2024 Microsoft and its suppliers. All rights reserved.

43.70567, -72.30625


Route I91 /  VT10A

Structure #0048N /  0001A(Routine)

I 00091 ML over I 91 OVER VT 10A

Team Lead: Stephen Piro,  Inspection Date: 04/15/2024

IDENTIFICATION

CLASSIFICATION

(1) State Names
(8) Structure Number
(5) Inventory Route
(2) Highway Agency District
(3) County Code
(4) Place Code
(6) Features Intersected
(7) Facility Carried
(9) Location
(11) Mile Point
(12) Base Highway Network
(13) LRS Inventory Rte & Subrte
(16) Latitude
(17) Longitude

### Get maintenance section boundaries

In [1555]:
def find_maintenance_sections(doc):
    """
    Find all "Maintenance Needs" sections in the document, from the header to "Remarks".
    
    Args:
        doc: spaCy Doc object
        
    Returns:
        list: List of section text strings
    """
    text = doc.text
    
    # Pattern to match "Maintenance Needs" header
    maintenance_pattern = r'Maintenance\s+Needs'
    
    # Pattern to match "Remarks" header
    remarks_pattern = r'Remarks'
    
    # Find all occurrences of "Maintenance Needs"
    maintenance_matches = list(re.finditer(maintenance_pattern, text, re.IGNORECASE))
    
    sections = []
    
    for i, match in enumerate(maintenance_matches):
        # Find the start of this maintenance section
        section_start = match.start()
        
        # Find the end of this maintenance section (next "Remarks" or next "Maintenance Needs" or end of text)
        # First look for "Remarks" after this match
        remarks_match = re.search(remarks_pattern, text[section_start:], re.IGNORECASE)
        if remarks_match:
            section_end = section_start + remarks_match.start()
        else:
            # If no "Remarks" found, check if there's another "Maintenance Needs"
            if i < len(maintenance_matches) - 1:
                section_end = maintenance_matches[i + 1].start()
            else:
                # If this is the last maintenance section, go to the end of the text
                section_end = len(text)
        
        # Extract the section text
        section_text = text[section_start:section_end].strip()
        sections.append(section_text)
    
    return sections

In [1556]:
def find_maintenance_sections(doc):
    """
    Find all "Maintenance Needs" sections in the document, from the header to "Remarks".
    
    Args:
        doc: spaCy Doc object
        
    Returns:
        list: List of section text strings
    """
    text = doc.text
    
    # Pattern to match "Maintenance Needs" header
    maintenance_pattern = r'Maintenance\s+Needs'
    
    # Pattern to match "Remarks" header
    remarks_pattern = r'Remarks'
    
    # Find all occurrences of "Maintenance Needs"
    maintenance_matches = list(re.finditer(maintenance_pattern, text, re.IGNORECASE))
    
    sections = []
    
    for i, match in enumerate(maintenance_matches):
        # Find the start of this maintenance section
        section_start = match.start()
        
        # Find the end of this maintenance section (next "Remarks" or next "Maintenance Needs" or end of text)
        # First look for "Remarks" after this match
        remarks_match = re.search(remarks_pattern, text[section_start:], re.IGNORECASE)
        
        if remarks_match:
            section_end = section_start + remarks_match.start()
        else:
            # If no "Remarks" found, check if there's another "Maintenance Needs"
            if i < len(maintenance_matches) - 1:
                section_end = maintenance_matches[i + 1].start()
            else:
                # If this is the last maintenance section, go to the end of the text
                section_end = len(text)
        
        # Extract the section text
        section_text = text[section_start:section_end].strip()
        sections.append(section_text)
    
    return sections

In [1557]:
maintenance_sections = find_maintenance_sections(doc4)
len(maintenance_sections)

3

In [1558]:
for mainten in maintenance_sections:
    print(mainten)

Maintenance Needs

Date Reported: 04/21/2022

Priority:

4 - Maintenance Finding - Next 
Inspection Cycle

Status:

Open

Type of Work:

30 - Substructure - Abutment patching or 
repair

Component:

Substructure

Deficiency Description

Reinforced concrete abutment #1 has minor to moderate cracking scattered throughout with heavy rust stains and 
efflorescence leakage along the stem.  Abutment #1 backwall has heavy spalling/ scaling present with highly saturated 
concrete at both the eastern and western end exposing the steel reinforcing.  Reinforced concrete abutment #2 has heavy 
spalling along the stem and bridge seat exposing the steel reinforcing with measurable section loss along the outer 
portions and multiple areas along the bridge seat with spalling and pooling water.
Maintenance Needs

Date Reported: 04/21/2022

Priority:

4 - Maintenance Finding - Next 
Inspection Cycle

Status:

Open

Type of Work:

31 - Substructure - Bearing seat repair

Component:

Substructure

Deficie

### Parse maintenance section boundaries

In [1559]:
def extract_maintenance_info(maintenance_sections: List[str]) -> List[Dict]:
    """
    Extract maintenance information from an array of text sections
    
    Args:
        maintenance_sections: List of text strings, each representing a maintenance section
        
    Returns:
        List of dictionaries containing extracted maintenance information
    """
    # List to store all processed maintenance sections
    processed_sections = []
    
    # Process each section
    for section_text in maintenance_sections:
        current_section = {}
        current_section["full_text"] = section_text
        # Extract date reported
        date_match = re.search(r"Date Reported:\s*(\d{1,2}/\d{1,2}/\d{4})", section_text)
        if date_match:
            current_section["date_reported"] = date_match.group(1).strip()
        
        # Extract priority - look for content between "Priority:" and the next field
        priority_match = re.search(r"Priority:(?:\s*\n)?(.*?)(?=\n\s*(?:Type of Work:|Status:|Component:))", section_text, re.DOTALL)
        if priority_match:
            priority_text = priority_match.group(1).strip()
            # Clean up multi-line text
            priority_text = re.sub(r'\s+', ' ', priority_text)
            current_section["priority"] = priority_text
        
        # Extract type of work - look for content between "Type of Work:" and the next field
        work_match = re.search(r"Type of Work:(?:\s*\n)?(.*?)(?=\n\s*(?:Status:|Component:|Deficiency))", section_text, re.DOTALL)
        if work_match:
            work_text = work_match.group(1).strip()
            # Clean up multi-line text
            work_text = re.sub(r'\s+', ' ', work_text)
            current_section["type_of_work"] = work_text
        
        # Extract status
        status_match = re.search(r"Status:(?:\s*\n)?(.*?)(?=\n\s*(?:Component:|Deficiency))", section_text, re.DOTALL)
        if status_match:
            status_text = status_match.group(1).strip()
            current_section["status"] = status_text
        
        # Extract component
        component_match = re.search(r"Component:(?:\s*\n)?(.*?)(?=\n\s*(?:Deficiency|Remarks))", section_text, re.DOTALL)
        if component_match:
            component_text = component_match.group(1).strip()
            current_section["component"] = component_text
        
        # Try multiple patterns for deficiency description, just like in the first code
        deficiency_patterns = [
            r"Deficiency Description\s*(.+?)(?=Remarks:|$)",
            r"Deficiency Description\n(.+?)(?=\n\n|\n[A-Z]|$)"
        ]
        
        for pattern in deficiency_patterns:
            deficiency_match = re.search(pattern, section_text, re.DOTALL | re.IGNORECASE)
            if deficiency_match:
                deficiency_text = deficiency_match.group(1).strip()
                # Clean up multi-line text
                deficiency_text = re.sub(r'\s+', ' ', deficiency_text)
                current_section["deficiency_description"] = deficiency_text
                break  # Found a match, no need to try other patterns
        
        # Only add sections that have at least some data
        if current_section:
            processed_sections.append(current_section)
    
    return processed_sections

In [1560]:
# print(maintenance_sections[0])

In [1561]:
# type(maintenance_sections[0])

In [1562]:
maintenance_sections

['Maintenance Needs\n\nDate Reported: 04/21/2022\n\nPriority:\n\n4 - Maintenance Finding - Next \nInspection Cycle\n\nStatus:\n\nOpen\n\nType of Work:\n\n30 - Substructure - Abutment patching or \nrepair\n\nComponent:\n\nSubstructure\n\nDeficiency Description\n\nReinforced concrete abutment #1 has minor to moderate cracking scattered throughout with heavy rust stains and \nefflorescence leakage along the stem.  Abutment #1 backwall has heavy spalling/ scaling present with highly saturated \nconcrete at both the eastern and western end exposing the steel reinforcing.  Reinforced concrete abutment #2 has heavy \nspalling along the stem and bridge seat exposing the steel reinforcing with measurable section loss along the outer \nportions and multiple areas along the bridge seat with spalling and pooling water.',
 'Maintenance Needs\n\nDate Reported: 04/21/2022\n\nPriority:\n\n4 - Maintenance Finding - Next \nInspection Cycle\n\nStatus:\n\nOpen\n\nType of Work:\n\n31 - Substructure - Beari

In [1563]:
mainntenance_needs = extract_maintenance_info(maintenance_sections)
mainntenance_needs

[{'full_text': 'Maintenance Needs\n\nDate Reported: 04/21/2022\n\nPriority:\n\n4 - Maintenance Finding - Next \nInspection Cycle\n\nStatus:\n\nOpen\n\nType of Work:\n\n30 - Substructure - Abutment patching or \nrepair\n\nComponent:\n\nSubstructure\n\nDeficiency Description\n\nReinforced concrete abutment #1 has minor to moderate cracking scattered throughout with heavy rust stains and \nefflorescence leakage along the stem.  Abutment #1 backwall has heavy spalling/ scaling present with highly saturated \nconcrete at both the eastern and western end exposing the steel reinforcing.  Reinforced concrete abutment #2 has heavy \nspalling along the stem and bridge seat exposing the steel reinforcing with measurable section loss along the outer \nportions and multiple areas along the bridge seat with spalling and pooling water.',
  'date_reported': '04/21/2022',
  'priority': '4 - Maintenance Finding - Next Inspection Cycle',
  'type_of_work': '30 - Substructure - Abutment patching or repair'

## Component and maintenance needs matching

**Important** -- heabily double check this part you made some changes to the extractor and to the maintenance needs appender

In [1564]:
updated_components_new = []
for component in updated_components:
        updated_component = component.copy()
        updated_component["maintenance_needs"] = []
        updated_components_new.append(updated_component)
updated_components_new

[{'text': '205 Reinforced Concrete Column',
  'full_section': '205 Reinforced Concrete Column EA 6 6 0 0 0',
  'defects': [],
  'id': '205',
  'units': 'EA',
  'values': [6, 6, 0, 0, 0],
  'description_section': {'header': '85 - Pier Columns Condition',
   'condition_text': 'Good',
   'condition_rating': 'GOOD',
   'assessment': None,
   'content': 'Both piers have three (3) concrete columns that are in fairly good condition.'},
  'maintenance_needs': []},
 {'text': '215 Reinforced Concrete Abutment',
  'full_section': '215 Reinforced Concrete Abutment LF 92 0 12 60 20\n1080 Delamination/Spall/Patched Area LF 72 0 12 40 20\n1130 Cracking (RC and Other) LF 20 0 0 20 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 72,
    'cs_values': [0, 12, 40, 20]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'LF',
    'total': 20,
    'cs_values': [0, 0, 20, 0]}],
  'id': '215',
  'units': 'LF',
  'v

In [1565]:
def match_by_component_name(need, components):
    """
    Match a maintenance need to a component by direct component name comparison.
    
    Args:
        need (dict): Maintenance need dictionary
        components (list): List of component dictionaries
        
    Returns:
        dict or None: Matched component or None if no match found
    """
    if not need.get("component"):
        return None
    
    need_component = need["component"].lower()
    
    for component in components:
        # Extract clean component name
        clean_name = extract_component_name(component["text"]).lower()
        
        # Check for direct match or substring match
        if (clean_name == need_component or 
            clean_name in need_component or 
            need_component in clean_name):
            return component
        
        # Handle special cases
        if clean_name == "wing/retaining wall" and ("wing" in need_component or "retaining" in need_component or "wall" in need_component):
            return component
        
        if clean_name == "abutment" and "abutment" in need_component:
            return component
    
    return None

In [1566]:
def match_by_description(need, components):
    """
    Match a maintenance need to a component by finding component mentions in the description.
    
    Args:
        need (dict): Maintenance need dictionary
        components (list): List of component dictionaries
        
    Returns:
        dict or None: Matched component or None if no match found
    """
    if not need.get("deficiency_description"):
        return None
    
    description = need["deficiency_description"].lower()
    
    # Try each component
    for component in components:
        print(component["text"])
        print(extract_component_name(component["text"]))
        # Extract clean component name and variations
        clean_name = extract_component_name(component["text"])
        variations = generate_component_variations(clean_name)
        print(variations)
        
        # Check if any variation appears in the description
        for variation in variations:
            pattern = r'\b' + variation.lower() + r'\b'
            if re.search(pattern, description):
                return component
            
            # Handle numbered components like "abutment 1"
            if variation.lower() + r' \d+' in variations:
                pattern = r'\b' + variation.lower() + r' \d+\b'
                if re.search(pattern, description):
                    return component
    
    return None


In [1567]:
import re

def match_by_description(need, components):
    """
    Match a maintenance need to a component by finding component mentions in the description.
    
    Args:
        need (dict): Maintenance need dictionary
        components (list): List of component dictionaries
        
    Returns:
        dict or None: Matched component or None if no match found
    """
    if not need.get("deficiency_description"):
        return None
    
    description = need["deficiency_description"].lower()
    
    # Special case for bearings
    bearing_matches = []
    if "bearing" in description or "bearings" in description:
        for component in components:
            if "bearing" in component["text"].lower():
                bearing_matches.append(component)
        
        # If we found multiple bearing components, prioritize based on context
        if len(bearing_matches) > 1:
            # Check for specific bearing types in description
            if "fixed" in description and any("fixed" in c["text"].lower() for c in bearing_matches):
                for comp in bearing_matches:
                    if "fixed" in comp["text"].lower():
                        return comp
            elif "movable" in description and any("movable" in c["text"].lower() for c in bearing_matches):
                for comp in bearing_matches:
                    if "movable" in comp["text"].lower():
                        return comp
            
            # If description mentions bearing numbers, check both components
            if re.search(r'bearing\s+#?\d+', description, re.IGNORECASE):
                # Here's where we modify behavior - we add maintenance need to ALL bearing components
                # but still return only the first one to maintain compatibility
                
                # First, add the maintenance need to ALL bearing components
                for comp in bearing_matches:
                    # Get the original need to avoid reference issues
                    comp["maintenance_needs"].append(need)
                
                # Return the first one to maintain compatibility with existing code
                # (Note: we've already added the need to all bearing components)
                return bearing_matches[0]
    
    # Try each component
    for component in components:
        # Extract clean component name and variations
        clean_name = extract_component_name(component["text"])
        variations = generate_component_variations(clean_name)
        
        # Check if any variation appears in the description
        for variation in variations:
            pattern = r'\b' + variation.lower() + r'\b'
            if re.search(pattern, description):
                return component
            
            # Handle numbered components like "abutment 1"
            if variation.lower() + r' \d+' in variations:
                pattern = r'\b' + variation.lower() + r' \d+\b'
                if re.search(pattern, description):
                    return component
    
    return None


In [1568]:
# Process each maintenance need

for need in mainntenance_needs:
            # Try to match by direct component name
    matched_component = match_by_component_name(need, updated_components_new)
            
            # If no match, try to match by mentions in the deficiency description
    if not matched_component:
        matched_component = match_by_description(need, updated_components_new)

        # If a match was found, add the maintenance need to the component
    if matched_component:
        if len(matched_component["maintenance_needs"]) < 1:
                matched_component["maintenance_needs"].append(need)
        
updated_components_new


[{'text': '205 Reinforced Concrete Column',
  'full_section': '205 Reinforced Concrete Column EA 6 6 0 0 0',
  'defects': [],
  'id': '205',
  'units': 'EA',
  'values': [6, 6, 0, 0, 0],
  'description_section': {'header': '85 - Pier Columns Condition',
   'condition_text': 'Good',
   'condition_rating': 'GOOD',
   'assessment': None,
   'content': 'Both piers have three (3) concrete columns that are in fairly good condition.'},
  'maintenance_needs': []},
 {'text': '215 Reinforced Concrete Abutment',
  'full_section': '215 Reinforced Concrete Abutment LF 92 0 12 60 20\n1080 Delamination/Spall/Patched Area LF 72 0 12 40 20\n1130 Cracking (RC and Other) LF 20 0 0 20 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 72,
    'cs_values': [0, 12, 40, 20]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'LF',
    'total': 20,
    'cs_values': [0, 0, 20, 0]}],
  'id': '215',
  'units': 'LF',
  'v

## Add bridge id context and general observation

In [1569]:
go = sections[5]["content"].split("\n")
go = "\n".join(go[:go.index("")])
print(go)

GENERAL OBSERVATION


In [1570]:
go.split("/n")

['GENERAL OBSERVATION']

In [1571]:
for com in updated_components_new:
    com["bridge_id"] = bridge_id
    if len(go.split("\n")) > 1:
        com["general_observation"] = go
updated_components_new

[{'text': '205 Reinforced Concrete Column',
  'full_section': '205 Reinforced Concrete Column EA 6 6 0 0 0',
  'defects': [],
  'id': '205',
  'units': 'EA',
  'values': [6, 6, 0, 0, 0],
  'description_section': {'header': '85 - Pier Columns Condition',
   'condition_text': 'Good',
   'condition_rating': 'GOOD',
   'assessment': None,
   'content': 'Both piers have three (3) concrete columns that are in fairly good condition.'},
  'maintenance_needs': [],
  'bridge_id': '0048'},
 {'text': '215 Reinforced Concrete Abutment',
  'full_section': '215 Reinforced Concrete Abutment LF 92 0 12 60 20\n1080 Delamination/Spall/Patched Area LF 72 0 12 40 20\n1130 Cracking (RC and Other) LF 20 0 0 20 0',
  'defects': [{'id': '1080',
    'description': 'Delamination/Spall/Patched Area',
    'unit': 'LF',
    'total': 72,
    'cs_values': [0, 12, 40, 20]},
   {'id': '1130',
    'description': 'Cracking (RC and Other)',
    'unit': 'LF',
    'total': 20,
    'cs_values': [0, 0, 20, 0]}],
  'id': '215'

## You can do maintenacne need thing here